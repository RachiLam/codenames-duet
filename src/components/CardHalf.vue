<template>
  <div
    class="card-half"
    :class="[card.status[playerId], position]"
    @click="toggleButtons"
  >
    <div class="word">
      {{ card.word }}
    </div>
    <div
      v-if="showButtons && adminMode"
      class="button-group"
    >
      <button
        class="reroll"
        @click="setStatus('reroll')"
      >
        REROLL
      </button>
      <button
        class="disallow"
        @click="setStatus('disallow')"
      >
        DISALLOW
      </button>
    </div>
    <div
      v-if="showButtons && !adminMode"
      class="button-group"
    >
      <button
        class="reset"
        @click="setStatus(null)"
      />
      <button
        class="assassin"
        @click="setStatus('assassin')"
      />
      <button
        class="bystander"
        @click="setStatus('bystander')"
      />
      <button
        class="agent"
        @click="setStatus('agent')"
      />
    </div>
  </div>
</template>

<script>
import { mapMutations, mapState } from 'vuex';

export default {
  props: {
    card: {
      type: Object,
      required: true,
    },
    playerId: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      showButtons: false,
    };
  },
  computed: {
    ...mapState(['adminMode']),
    position() {
      return this.playerId === 0 ? 'top' : 'bottom';
    },
  },
  methods: {
    ...mapMutations(['updateCardStatus']),
    toggleButtons() {
      this.showButtons = !this.showButtons;
    },
    setStatus(status) {
      this.updateCardStatus({
        cardId: this.card.id,
        playerId: this.playerId,
        status,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
  @use 'styles/colors';
  @use 'styles/padding';

  .card-half {
    display: flex;
    flex-direction: column;
    flex: 1;
    flex-shrink: 0;
    border-radius: 5px 5px 0px 0px;
    padding: padding.$md;
    padding-bottom: padding.$sm;
    font-size: 25px;

    &.bottom {
      transform: rotate(-180deg);
      font-style: italic;
    }

    &.agent,
    &.assassin,
    &.bystander {
      border-bottom: 0.5px solid colors.$accent-light;

      * {
        color: colors.$accent-light;
      }
    }

    &.agent {
      background-color: colors.$success;
    }

    &.assassin {
      background-color: colors.$assassin;
    }

    &.bystander {
      background-color: colors.$accent-dark;
    }

    .button-group {
      display: flex;

      $size: 20px;
      button {
        flex: 1;
        min-width: $size;
        height: $size;
        border-radius: 10%;
        border: 1px solid colors.$accent-light;

        &:not(:last-of-type) {
          margin-right: padding.$sm;
        }

        &.agent {
          background-color: colors.$success;
        }

        &.bystander {
          background-color: colors.$accent-dark;
        }

        &.assassin {
          background-color: colors.$assassin;
        }

        &.reset, &.reroll, &.disallow {
          background-color: colors.$accent-light;
          border-color: colors.$accent-dark;
        }
      }
    }
  }
</style>
