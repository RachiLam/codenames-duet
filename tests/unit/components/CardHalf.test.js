import Vue from 'vue';
import CardHalf from 'components/CardHalf.vue';
import { genCard } from 'helpers/generators';

const { shallowMount } = getMountUtils({
  component: CardHalf,
  defaultProps: {
    card: genCard(),
    playerId: 0,
  },
});

const getWordContainer = (wrapper) => wrapper.find('.word');
const getButtonGroup = (wrapper) => wrapper.find('.button-group');

describe('components/CardHalf', () => {
  context('by default', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount({
        props: {
          card: genCard({ word: 'THE WORD' }),
        },
      });
    });

    it('renders the word', () => {
      expect(getWordContainer(wrapper).element).toMatchSnapshot();
    });

    it('does not render buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });

  context('when player id is 0', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount({
        props: {
          playerId: 0,
        },
      });
    });

    it('renders as the top word', () => {
      expect(wrapper.classes()).toContain('top');
    });
  });

  context('when player id is 1', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount({
        props: {
          playerId: 1,
        },
      });
    });

    it('renders as the bottom word', () => {
      expect(wrapper.classes()).toContain('bottom');
    });
  });

  context('when pressed', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount();

      wrapper.trigger('click');
    });

    it('renders buttons', () => {
      expect(getButtonGroup(wrapper).element).toMatchSnapshot();
    });
  });

  context('when pressed twice', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount();

      wrapper.trigger('click');
      wrapper.trigger('click');
    });

    it('hides the buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });

  context('when the agent button is pressed', () => {
    let wrapper;
    const updateCardStatus = jest.fn();
    beforeAll((done) => {
      wrapper = shallowMount({
        props: {
          card: genCard({ id: 4 }),
          playerId: 1,
        },
        flatStore: {
          updateCardStatus,
        },
      });

      wrapper.trigger('click');
      Vue.nextTick(() => {
        getButtonGroup(wrapper).find('.agent').trigger('click');
        done();
      });
    });

    it('sets the cards status', () => {
      expect(updateCardStatus).toHaveBeenCalledWith(
        expect.anything(),
        { cardId: 4, playerId: 1, status: 'agent' },
      );
    });

    it('hides the buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });

  context('when the bystander button is pressed', () => {
    let wrapper;
    const updateCardStatus = jest.fn();
    beforeAll((done) => {
      wrapper = shallowMount({
        props: {
          card: genCard({ id: 7 }),
          playerId: 0,
        },
        flatStore: {
          updateCardStatus,
        },
      });

      wrapper.trigger('click');
      Vue.nextTick(() => {
        getButtonGroup(wrapper).find('.bystander').trigger('click');
        done();
      });
    });

    it('sets the cards status', () => {
      expect(updateCardStatus).toHaveBeenCalledWith(
        expect.anything(),
        { cardId: 7, playerId: 0, status: 'bystander' },
      );
    });

    it('hides the buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });

  context('when the assassin button is pressed', () => {
    let wrapper;
    const updateCardStatus = jest.fn();
    beforeAll((done) => {
      wrapper = shallowMount({
        props: {
          card: genCard({ id: 2 }),
          playerId: 1,
        },
        flatStore: {
          updateCardStatus,
        },
      });

      wrapper.trigger('click');
      Vue.nextTick(() => {
        getButtonGroup(wrapper).find('.assassin').trigger('click');
        done();
      });
    });

    it('sets the cards status', () => {
      expect(updateCardStatus).toHaveBeenCalledWith(
        expect.anything(),
        { cardId: 2, playerId: 1, status: 'assassin' },
      );
    });

    it('hides the buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });

  context('when the reset button is pressed', () => {
    let wrapper;
    const updateCardStatus = jest.fn();
    beforeAll((done) => {
      wrapper = shallowMount({
        props: {
          card: genCard({ id: 6 }),
          playerId: 0,
        },
        flatStore: {
          updateCardStatus,
        },
      });

      wrapper.trigger('click');
      Vue.nextTick(() => {
        getButtonGroup(wrapper).find('.reset').trigger('click');
        done();
      });
    });

    it('sets the cards status', () => {
      expect(updateCardStatus).toHaveBeenCalledWith(
        expect.anything(),
        { cardId: 6, playerId: 0, status: null },
      );
    });

    it('hides the buttons', () => {
      expect(getButtonGroup(wrapper).exists()).toBe(false);
    });
  });
});
