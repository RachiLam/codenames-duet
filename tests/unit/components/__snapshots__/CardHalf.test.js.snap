// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CardHalf by default renders the word 1`] = `
<div
  class="word"
>
  
    THE WORD
  
</div>
`;

exports[`components/CardHalf when pressed renders buttons 1`] = `
<div
  class="button-group"
>
  <button
    class="reset"
  />
   
  <button
    class="assassin"
  />
   
  <button
    class="bystander"
  />
   
  <button
    class="agent"
  />
</div>
`;
